token-lending/program/src/processor.rs

```diff
@@ -578,6 +587,309 @@ fn _deposit_reserve_liquidity<'a>(
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &lending_market.token_program_id != token_program_id.key {
+        msg!("Lending market token program does not match the token program provided");
+        return Err(LendingError::InvalidTokenProgram.into());
+    }
+    let mut reserve = Reserve::unpack(&reserve_info.data.borrow())?;
+    if reserve_info.owner != program_id {
+        msg!("Reserve provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &reserve.lending_market != lending_market_info.key {
+        msg!("Reserve lending market does not match the lending market provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &reserve.liquidity.supply_pubkey != source_liquidity_info.key {
+        msg!("Borrow reserve liquidity supply must be used as the source liquidity provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &reserve.liquidity.supply_pubkey == destination_liquidity_info.key {
+        msg!(
+            "Borrow reserve liquidity supply cannot be used as the destination liquidity provided"
+        );
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if reserve.borrowing {
+        msg!("Multiple flash borrows not allowed");
+        return Err(LendingError::MultipleFlashBorrows.into());
+    }
+
+    let authority_signer_seeds = &[
+        lending_market_info.key.as_ref(),
+        &[lending_market.bump_seed],
+    ];
+    let lending_market_authority_pubkey =
+        Pubkey::create_program_address(authority_signer_seeds, program_id)?;
+    if &lending_market_authority_pubkey != lending_market_authority_info.key {
+        msg!(
+            "Derived lending market authority {} does not match the lending market authority provided {}",
+            &lending_market_authority_pubkey.to_string(),
+            &lending_market_authority_info.key.to_string(),
+        );
+        return Err(LendingError::InvalidMarketAuthority.into());
+    }
+
+    // @FIXME: if u64::MAX is flash loaned, fees should be inclusive as with ordinary borrows
+    let flash_loan_amount = if liquidity_amount == u64::MAX {
+        reserve.liquidity.available_amount
+    } else {
+        liquidity_amount
+    };
+
+    let flash_loan_amount_decimal = Decimal::from(flash_loan_amount);
+
+    // Make sure this isnt a cpi call
+    let current_index = load_current_index_checked(&sysvar_info)? as usize;
+    let current_ixn = load_instruction_at_checked(current_index, &sysvar_info)?;
+    if current_ixn.program_id != *program_id {
+        msg!("Cpi call found: Flash borrow from {} does not match program_id {} ", current_ixn.program_id, *program_id);
+        return Err(LendingError::FlashBorrowCpi.into());
+    }
+
+    // TODO can we use the enum instead?
+    static FLASH_REPAY_INSTRUCTION: u8 = 18;
```

Comment:
Is it possible to get a reference to an Enum value's index? It would be nice if I could just use `LendingInstruction::FlashRepayReserveLiquidity` somehow instead of defining another constant.

Reply:
I don't think it's possible unless the Enum value is a primitive type. Even so – we shouldn't rely on this because this logic is contained within `LendingInstruction::unpack`.

Reply:
We could modify `LendingInstruction::unpack` to use enums that are defined somewhere else, and then use those enums here.